script/foreman-debug

```diff
@@ -79,50 +79,79 @@ add_cmd() {
-      printv " - $FILE"
-      SUBDIR=$(dirname $FILE)
-      [ ! -d "$DIR$SUBDIR" ] && mkdir -p "$DIR$SUBDIR"
-      MIME=$(file -bi "$FILE" | cut -d\; -f1)
-      case $MIME in
+# add and filter file of known MIME type, calculate size
+add_file() {
+  FILE=$1
+  SIZE=$(du -k $FILE | cut -f1)
+  MIME=$(file -bi "$FILE" | cut -d\; -f1)
+  case $MIME in
         application/x-gzip)
-          zcat "$FILE" | $TAIL_OR_CAT | sed -r "$FILTER" > "$DIR$FILE.txt"
+          zcat "$FILE" | sed -r "$FILTER" > "$DIR$FILE.txt"
           ;;
         application/x-bzip2)
-          bzcat "$FILE" | $TAIL_OR_CAT | sed -r "$FILTER" > "$DIR$FILE.txt"
+          bzcat "$FILE" | sed -r "$FILTER" > "$DIR$FILE.txt"
           ;;
         application/x-xz)
-          xzcat "$FILE" | $TAIL_OR_CAT | sed -r "$FILTER" > "$DIR$FILE.txt"
+          xzcat "$FILE" | sed -r "$FILTER" > "$DIR$FILE.txt"
           ;;
         text/plain | application/xml)
-          $TAIL_OR_CAT "$FILE" | sed -r "$FILTER" > "$DIR$FILE"
+          sed -r "$FILTER" "$FILE" > "$DIR$FILE"
           [ $PRINTPASS -eq 1 ] && grep -H "\*\*\*\*\*" "$DIR$FILE"
           ;;
-        inode/symlink)
-          cp -a "$FILE" "$DIR$FILE"
-          add_files $(readlink -f $FILE)
-          ;;
         *)
           echo "Skipping file $FILE: unknown MIME type $MIME" >> "$DIR/skipped_files"
+          SIZE=0 # don't count the size to collected files
           ;;
       esac
+}
+
+# add files (from newest to oldest) that are non zero, readable, regular file or symlink, until $MAXSIZE limit is reached
+add_files() {
+  SUMSIZE=0
+  SKIP="" # will be nonempty to stop adding files
+  # sort files per modification time, newest first
+  for FILE in $(ls -t $* 2> /dev/null); do
+    if [ \( -f "$FILE" -o -h "$FILE" \) -a \( -r "$FILE" -a -s "$FILE" \) ]; then
+      printv " - $FILE$SKIP"
+      if [ -z "$SKIP" ]; then
+        SUBDIR=$(dirname $FILE)
+        [ ! -d "$DIR$SUBDIR" ] && mkdir -p "$DIR$SUBDIR"
+        # if the file is symlink, copy source and add target
+        if [ -h $FILE ]; then
+          cp -a "$FILE" "$DIR$FILE"
+          FILE=$(readlink -f $FILE)
+        fi
+        add_file $FILE
+        SUMSIZE=$(($((SUMSIZE))+$((SIZE))))
+        # if we are over the size limit, skip rest older files
+        if [ \( $SUMSIZE -gt $MAXSIZE \) -a \( $MAXSIZE -gt 0 \) ]; then
+          SKIP=" (skipped due to size)"
+        fi
+      fi
     fi
   done
 }
 
+# call add_files for groups of files that have unique prefix separated by 1st argument
+# example: add_files_types "[.]log" /var/log/foreman/*.log* will call add_files to:
+# /var/log/foreman/apipe_cache.log*
+# /var/log/foreman/cron.log*
+# etc.
+# be aware, 1st argument is regexp, so use "[.]" instead of "." for dot char
+add_files_types() {
+  SEPARATOR="$1"
+  shift
+  FILES="$*"
+  for FILETYPE in $(ls $FILES | awk -F "$SEPARATOR" -v sep="$SEPARATOR" '{ print $1 sep }' | sort -u); do
+    add_files "${FILETYPE}*"
+  done
+}
```

Comment:
Again, I'd strongly prefer GNU find. More importantly, I don't like using regexp patterns, globs are very expected in this case. Find can do the processing for you (it can do globs). Do you have a use case where you need to have regexp instead of a glob?